File: backend/app/schemas/analytics.py
"""Pydantic schemas for analytics events and content analytics query API."""

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class AnalyticsEventCreate(BaseModel):
    """Request body for POST /api/v1/analytics/events."""

    event_type: Literal["play_start", "play_pause", "play_complete", "browse", "search"]
    title_id: uuid.UUID | None = None
    service_type: Literal["Linear", "VoD", "SVoD", "TSTV", "Catch_up", "Cloud_PVR"]
    profile_id: uuid.UUID | None = None
    region: str
    occurred_at: datetime
    session_id: uuid.UUID | None = None
    duration_seconds: int | None = None
    watch_percentage: int | None = Field(None, ge=0, le=100)
    extra_data: dict | None = None


class QueryRequest(BaseModel):
    """Request body for POST /api/v1/content-analytics/query."""

    question: str = Field(..., min_length=3, max_length=1000)


class QueryResult(BaseModel):
    """Structured result returned by the query engine."""

    summary: str
    confidence: float = Field(ge=0.0, le=1.0)
    data: list[dict]
    applied_filters: dict
    data_sources: list[str]
    data_freshness: datetime
    coverage_start: datetime


class ClarificationResponse(BaseModel):
    """Response when the agent cannot determine query intent."""

    type: Literal["clarification"] = "clarification"
    clarifying_question: str
    context: str


class QueryResponse(BaseModel):
    """Union response for POST /api/v1/content-analytics/query."""

    status: Literal["complete", "pending", "clarification"]
    result: QueryResult | None = None
    job_id: uuid.UUID | None = None
    clarification: ClarificationResponse | None = None


class JobStatusResponse(BaseModel):
    """Response for GET /api/v1/content-analytics/jobs/{job_id}."""

    job_id: uuid.UUID
    status: Literal["pending", "complete", "failed"]
    submitted_at: datetime
    completed_at: datetime | None = None
    result: QueryResult | None = None
    error_message: str | None = None

File: frontend-admin/src/api/analytics.ts
import { apiFetch } from '@/api/client'

// ── Types (T003) ──────────────────────────────────────────────────────────────

export interface QueryRequest {
  question: string // 3–1000 characters
}

export interface QueryResult {
  summary: string
  confidence: number // 0.0–1.0
  data: Record<string, unknown>[]
  applied_filters: {
    region: string | null
    time_period: string | null
    service_type: string | null
  }
  data_sources: string[]
  data_freshness: string // ISO 8601 datetime
  coverage_start: string // ISO 8601 datetime
}

export interface ClarificationPayload {
  type: 'clarification'
  clarifying_question: string
  context: string
}

export interface QueryResponse {
  status: 'complete' | 'pending' | 'clarification'
  result: QueryResult | null
  job_id: string | null
  clarification: ClarificationPayload | null
}

export interface JobStatusResponse {
  job_id: string
  status: 'pending' | 'complete' | 'failed'
  submitted_at: string // ISO 8601
  completed_at: string | null
  result: QueryResult | null
  error_message: string | null
}

export interface HistoryEntry {
  id: string // uuid generated client-side for React key
  question: string
  submittedAt: Date
  status: 'complete' | 'clarification' | 'failed'
  result: QueryResult | null
  clarification: ClarificationPayload | null
  errorMessage: string | null
}

export type ActiveQueryState =
  | { phase: 'idle' }
  | { phase: 'submitting' }
  | { phase: 'polling'; jobId: string }
  | { phase: 'complete'; result: QueryResult }
  | { phase: 'clarification'; payload: ClarificationPayload }
  | { phase: 'failed'; errorMessage: string }
  | { phase: 'timeout' }

// ── API functions ─────────────────────────────────────────────────────────────

// T004
export function submitQuery(question: string): Promise<QueryResponse> {
  return apiFetch<QueryResponse>('/content-analytics/query', {
    method: 'POST',
    body: JSON.stringify({ question }),
  })
}

// T005
export function pollJob(jobId: string): Promise<JobStatusResponse> {
  return apiFetch<JobStatusResponse>(`/content-analytics/jobs/${jobId}`)
}
